packages/contracts/contracts/UnipigTransitionEvaluator.sol

```diff
@@ -0,0 +1,331 @@
+        dt.SwapTx memory swapTx = dt.SwapTx(
+            sender,
+            _transition.tokenType,
+            _transition.inputAmount,
+            _transition.minOutputAmount,
+            _transition.timeout
+        );
+
+        // Make sure that the provided storage slots are corrent
+        require(verifyEcdsaSignature(_transition.signature, getSwapTxHash(swapTx), sender), "Swap signature is invalid!");
+        require(_storageSlots[1].slotIndex == UNISWAP_SLOT_INDEX && recipient == UNISWAP_ADDRESS, "Swap tx must be swapping with Unipig!");
+
+        // Create an array to store our output storage slots
+        dt.Storage[2] memory outputStorage;
+
+        // Now we know the storage slots are correct, let's first make sure the sender has enough money to initiate the swap
+        uint senderBalance = _storageSlots[0].value.balances[swapTx.tokenType];
+        // Make sure the sender has enough money
+        require(senderBalance > swapTx.inputAmount, "Sender of the swap tx does not have enough money!");
+
+        // Store variables used for calculating the SWAP
+        uint inputTokenType = swapTx.tokenType;
+        uint outputTokenType = 1 - swapTx.tokenType;
+        dt.Storage memory senderStorage = _storageSlots[0].value;
+        dt.Storage memory uniswapStorage = _storageSlots[1].value;
+
+        // Compute the SWAP
+        uint invariant = uniswapStorage.balances[0] * uniswapStorage.balances[1];
+        uint inputWithFee = swapTx.inputAmount * (10000 - UNISWAP_FEE_IN_BIPS) / 10000;
+        uint totalInput = inputWithFee + uniswapStorage.balances[inputTokenType];
+        uint newOutputBalance = invariant / totalInput;
+        uint32 outputAmount = uniswapStorage.balances[outputTokenType] - uint32(newOutputBalance);
+        // Make sure the output amount is above or equal to the minimum
+        require(outputAmount >= swapTx.minOutputAmount, "Swap output amount is not above or equal to min output!");
+
+        // Update the sender storage slots with the new balances
+        senderStorage.balances[inputTokenType] -= swapTx.inputAmount;
+        senderStorage.balances[outputTokenType] += outputAmount;
+        // Update uniswap storage slots with the new balances
+        uniswapStorage.balances[inputTokenType] += swapTx.inputAmount;
+        uniswapStorage.balances[outputTokenType] -= outputAmount;
+
+        // Set our output storage
+        outputStorage[0] = senderStorage;
+        outputStorage[1] = uniswapStorage;
+        // Return the outputs!
+        return outputStorage;
+    }
+
+    /**
+     * Get the hash of the storage value.
+     */
+    function getStorageHash(dt.Storage memory _storage) public pure returns(bytes32) {
+        // Here we don't use `abi.encode([struct])` because it's not clear
+        // how to generate that encoding client-side.
+        return keccak256(abi.encode(_storage.pubkey, _storage.balances[0], _storage.balances[1]));
+    }
+
+    /************
+     * Decoding *
```

Comment:
*Important* `tokenType` was (and still is) listed as a `bool` in all our client encoding/decoding, whereas here it's listed as `uint`.

Reply:
Oh so as it turns out, bool encoding is actually identical to `uint`, so that's not a big problem. That said, I saw this is changed in the new encoding which this uses now so hopefully not an issue anymore. 

One note somewhat related to this is that I am currently doing a very dumb manual conversion from `balances` being an object with keys of '0', & '1', into `balances` being an array where the first item is uni and the 2nd is pigi. You can see it here: https://github.com/plasma-group/pigi/pull/448/files#diff-364498878b2bb730b65a6a87f059e45aR475

This is pretty annoying unfortunately.

Reply:
Will this require a change on the client? It should be easy to update [de-]serialization so that we sign it the way you want it but also represent it in JS how we want.

Reply:
This is a weird thing where the way you *have* to pass it in to ethers.js is an array. This is because they do the struct serialization on their end and so they need it in an object which is represented the same way that it is serialized. Additionally, you can't have mappings in structs, so there's no way to represent normal objects. This means that we've got to do this somewhat ugly transformation, or on the client side also use an array.